File: app/src/main/java/com/task/temponewstask/ui/adapter/NewsAdapter.java
package com.task.temponewstask.ui.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.task.temponewstask.R;
import com.task.temponewstask.model.dto.response.Articles;
import com.task.temponewstask.utils.TimeHelper;

import java.util.List;


public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.NewsViewHolder> {

    private List<Articles> articles;
    Context context;
    INewsAction iNewsAction;

    public NewsAdapter(List<Articles> articles, Context context, INewsAction iNewsAction) {
        this.articles = articles;
        this.context = context;
        this.iNewsAction = iNewsAction;
    }

    @Override
    public NewsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.viewholder_article, parent, false);

        return new NewsViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsViewHolder holder, int position) {
        Articles article = this.articles.get(position);
        holder.tv_article_title.setText(article.getTitle());
        holder.tv_article_author.setText(article.getAuthor());
        holder.tv_article_date.setText(TimeHelper.reformatBackEndDate(article.getPublishedAt()));
        holder.tv_article_description.setText(article.getDescription());
        holder.tv_article_source_text.setText(article.getSource().getName());
        holder.tv_article_title.setText(article.getTitle());
        holder.itemView.setOnClickListener(view -> {
            iNewsAction.clicked(article);
        });
        Glide.with(context)  //2
                .load(article.getUrlToImage()) //3
//                .centerCrop() //4
//                .placeholder(R.drawable.ic_image_place_holder) //5
//                .error(R.drawable.ic_broken_image) //6
//                .fallback(R.drawable.ic_no_image) //7
                .into(holder.iv_article_image);
        holder.btn_source.setOnClickListener((view) -> {
            Intent browse = new Intent(Intent.ACTION_VIEW, Uri.parse(article.getUrl()));
            context.startActivity(browse);

        });
    }

    @Override
    public int getItemCount() {
        return articles.size();
    }

    public void addItems(List<Articles> articles) {
        articles.addAll(articles);
        notifyDataSetChanged();
    }

    public class NewsViewHolder extends RecyclerView.ViewHolder {
        public TextView tv_article_title;
        public TextView tv_article_date;
        public TextView tv_article_description;
        public TextView tv_article_source_text;
        public TextView tv_article_author;
        public ImageView iv_article_image;
        public Button btn_source;

        public NewsViewHolder(View view) {
            super(view);
            tv_article_title = view.findViewById(R.id.tv_article_title);
            tv_article_date = view.findViewById(R.id.tv_article_date);
            tv_article_description = view.findViewById(R.id.tv_article_description);
            tv_article_source_text = view.findViewById(R.id.tv_article_source_text);
            tv_article_author = view.findViewById(R.id.tv_article_author);
            iv_article_image = view.findViewById(R.id.iv_article_image);
            btn_source = view.findViewById(R.id.btn_source);
        }

    }
}

File: app/src/main/java/com/task/temponewstask/ui/adapter/NewsAdapter.kt
package com.task.temponewstask.ui.adapter

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.task.temponewstask.R
import com.task.temponewstask.model.dto.response.Articles
import com.task.temponewstask.ui.adapter.NewsAdapter.NewsViewHolder
import com.task.temponewstask.utils.TimeHelper.reformatBackEndDate

class NewsAdapter(
    private val articles: ArrayList<Articles>,
    var context: Context,
    var iNewsAction: INewsAction
) : RecyclerView.Adapter<NewsViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.viewholder_article, parent, false)
        return NewsViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        val article = articles[position]
        holder.tv_article_title.text = article.title
        holder.tv_article_author.text = article.author
        holder.tv_article_date.text = reformatBackEndDate(article.publishedAt)
        holder.tv_article_description.text = article.description
        holder.tv_article_source_text.text = article.source.name
        holder.tv_article_title.text = article.title
        holder.itemView.setOnClickListener { view: View? ->
            iNewsAction.clicked(
                article
            )
        }
        Glide.with(context) //2
            .load(article.urlToImage) //3
            //                .centerCrop() //4
            //                .placeholder(R.drawable.ic_image_place_holder) //5
            //                .error(R.drawable.ic_broken_image) //6
            //                .fallback(R.drawable.ic_no_image) //7
            .into(holder.iv_article_image)
        holder.btn_source.setOnClickListener { view: View? ->
            val browse = Intent(Intent.ACTION_VIEW, Uri.parse(article.url))
            context.startActivity(browse)
        }
    }

    override fun getItemCount(): Int {
        return articles.size
    }

    fun addItems(articles: List<Articles>) {
        this.articles.addAll(articles)
        notifyDataSetChanged()
    }

    inner class NewsViewHolder(view: View) :
        RecyclerView.ViewHolder(view) {
        var tv_article_title: TextView
        var tv_article_date: TextView
        var tv_article_description: TextView
        var tv_article_source_text: TextView
        var tv_article_author: TextView
        var iv_article_image: ImageView
        var btn_source: Button

        init {
            tv_article_title = view.findViewById(R.id.tv_article_title)
            tv_article_date = view.findViewById(R.id.tv_article_date)
            tv_article_description = view.findViewById(R.id.tv_article_description)
            tv_article_source_text = view.findViewById(R.id.tv_article_source_text)
            tv_article_author = view.findViewById(R.id.tv_article_author)
            iv_article_image = view.findViewById(R.id.iv_article_image)
            btn_source = view.findViewById(R.id.btn_source)
        }
    }

}